### PR TITLE
feat: add juju 3.1 support

### PR DESCRIPTION
**Pull request by mert-kirpici**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -76,8 +77,12 @@ functional: build
 	@echo "Executing functional tests using built charm at ${PROJECTPATH}"
 	@CHARM_LOCATION=${PROJECTPATH} tox -e func -- ${FUNC_ARGS}
 
+functional31: build
+	@echo "Executing functional tests using built charm at ${PROJECTPATH} with juju 3.1 requirements"
+	@CHARM_LOCATION=${PROJECTPATH} tox -e func31 -- ${FUNC_ARGS}
```
Comment (rgildein) [inline Makefile:82]: `FUNC_ARGS` is not supported in your current implementation

File: tox.ini [hunk 2]
```diff
@@ -77,8 +77,23 @@ deps =
...
   git+https://github.com/openstack-charmers/zaza.git#egg=zaza
   -r {toxinidir}/tests/functional/requirements.txt
+
+[testenv:func31]
+setenv =
+  TEST_JUJU_CHANNEL = 3.1/stable
+changedir = {toxinidir}/tests/functional
+commands =
+  # NOTE(mertkirpici): functest-run-suite -b jammy focal
+  # or functest-run-suite -b jammy -b focal does not work
```
Comment (rgildein) [inline tox.ini:94]: I tried it and it's working fine for me to use `--bundle jammy focal`

File: tox.ini [hunk 3]
```diff
@@ -77,8 +77,23 @@ deps =
...
+  functest-run-suite --keep-faulty-model -b jammy
+  functest-run-suite --keep-faulty-model -b focal
```
Comment (rgildein) [inline tox.ini:96]:
```suggestion
  functest-run-suite --bundle jammy focal {posargs:--keep-faulty-model}
```
or maybe we can take advantage of aliases in zaza, so `juju_30: jammy` can be used so it's more obvious that which bundles are run on juju3.1.

**Review by rgildein — CHANGES_REQUESTED** [review]
I have also question about why we could not run functional test for 3.1 on bionic? Can you provide more information here. Thanks

File: tests/functional/tests/test_charm.py [hunk 4]
```diff
@@ -41,15 +41,15 @@ def setUp(self) -> None:
         self.unit = self.get_application_unit(self.NAME)
         self.controller = self.get_application_unit(self.CONTROLLER_APP)
 
-    def exec_juju_cmd(self, subcommand: str, raise_on_fail: bool = True) -> Dict[str, str]:
-        """Execute juju command on nested controller running within test environment.
+    def exec_cmd(self, command: str, raise_on_fail: bool = True) -> Dict[str, str]:
+        """Execute command on unit as the ubuntu user in the test environment.
```
[the remark this answers is not in the document]
Comment (mert-kirpici) [reply tests/functional/tests/test_charm.py:45]: we are running `juju wait` plugin, which works fine with 2.9 but, for the 3.1, the snap providing `juju-wait` binary is not in `juju` snap's `PATH` due to confinement, so it is not recognized as a juju plugin. Therefore we need to call it as `juju-wait` not `juju wait`. Therefore the method needs to be genericized.

**Comment by mert-kirpici** [comment]
> I have also question about why we could not run functional test for 3.1 on bionic? Can you provide more information here. Thanks

juju 3.1 requires the workload machine series to be greater or equal to focal.

File: tests/functional/tests/test_charm.py [hunk 5]
```diff
@@ -41,15 +41,15 @@ def setUp(self) -> None:
...
 
-        :param subcommand: Juju subcommand (add-model, deploy, ...) including params
+        :param command: Juju subcommand (add-model, deploy, ...) including params
```
Comment (agileshaw) [inline tests/functional/tests/test_charm.py:47]:
```suggestion
        :param command: command (juju add-model, juju deploy, juju-wait, ...) to run, including params
```

File: tox.ini (hunk 3 above)
Comment (agileshaw) [reply tox.ini:96]: +1 on adding back `postargs`

**Review by agileshaw — CHANGES_REQUESTED** [review]

File: tox.ini (hunk 2 above)
Comment (mert-kirpici) [reply tox.ini:94]: [Here](https://github.com/canonical/charm-prometheus-juju-exporter/actions/runs/6164621219/job/16730745388#step:9:251) is my trial run. If I read the output correctly, only the focal bundle was run.

`-b jammy -b focal` was a no go either.

Comment (rgildein) [reply tox.ini:94]: Can we report this as issue in zaza then? From `argparse` point of view it should work.

File: tox.ini (hunk 3 above)
Comment (rgildein) [reply tox.ini:96]:
```suggestion
  functest-run-suite -b jammy {posargs:--keep-faulty-model}
  functest-run-suite -b focal {posargs:--keep-faulty-model}
```

File: tox.ini (hunk 2 above)
Comment (mert-kirpici) [reply tox.ini:94]: yup. on it.

Comment (rgildein) [reply tox.ini:94]: Thanks :)

**Review by rgildein — APPROVED** [review]
LGTM

**Review by agileshaw — APPROVED** [review]
LGTM

**Pull request merged**